<a href="https://colab.research.google.com/github/akashdhara63/sales_data/blob/main/Copy_of_Assignmnet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df=pd.read_csv("/content/assessment_dataset.csv")

In [ ]:
df

TransactionID  CustomerID  TransactionDate  TransactionAmount  \
0                   1     16795.0    1/1/2022 0:00            1910.91   
1                   2      1860.0    1/1/2022 0:01           58590.27   
2                   3     39158.0    1/1/2022 0:02             280.21   
3                   4         NaN              NaN           86485.63   
4                   5     12284.0    1/1/2022 0:04             385.89   
...               ...         ...              ...                ...   
499995         499996     39050.0  12/14/2022 5:15             248.11   
499996         499997     20933.0  12/14/2022 5:16             852.55   
499997         499998     28355.0  12/14/2022 5:17             334.78   
499998         499999     42644.0  12/14/2022 5:18           99480.04   
499999         500000     35310.0  12/14/2022 5:19           33158.78   

       PaymentMethod  Quantity  DiscountPercent       City StoreType  \
0               Cash         1            35.00    Kolkata  In-Store   
1               Cash         1             9.51    Lucknow    Online   
2         Debit Card         1            16.55    Lucknow    Online   
3                UPI         1             9.53     Jaipur  In-Store   
4         Debit Card        22             0.98    Kolkata    Online   
...              ...       ...              ...        ...       ...   
499995   Credit Card        15            49.26  Ahmedabad    Online   
499996    Debit Card         3            44.61  Ahmedabad  In-Store   
499997    Debit Card        18            40.08     Jaipur    Online   
499998          Cash         1            32.99      Delhi    Online   
499999    Debit Card         1            23.95     Jaipur  In-Store   

        CustomerAge CustomerGender  LoyaltyPoints ProductName Region Returned  \
0              32.0           Male           2043     T-Shirt  North       No   
1              40.0          Other           8490        Sofa   West       No   
2              52.0          Other           6328    Notebook  North      Yes   
3              62.0          Other           1139      Laptop  North       No   
4              31.0         Female             66       Apple  North       No   
...             ...            ...            ...         ...    ...      ...   
499995         44.0           Male           2571       Apple  South      Yes   
499996         60.0         Female           2998     T-Shirt   East       No   
499997         65.0          Other           7476       Apple   East       No   
499998         34.0           Male           7815      Laptop  South      Yes   
499999         30.0           Male           4083        Sofa  North      Yes   

        FeedbackScore  ShippingCost  DeliveryTimeDays IsPromotional  
0                   4         24.31                 5            No  
1                   2       1502.96                 7           Yes  
2                   2         22.30                 2           Yes  
3                   4        461.38                 7            No  
4                   1         14.44                 1           Yes  
...               ...           ...               ...           ...  
499995              2         12.95                 1           Yes  
499996              2         24.77                 4            No  
499997              5          5.79                 3           Yes  
499998              2        513.26                 6            No  
499999              1       1763.72                 9            No  

[500000 rows x 19 columns]

In [ ]:
df.head(5)

TransactionID  CustomerID TransactionDate  TransactionAmount PaymentMethod  \
0              1     16795.0   1/1/2022 0:00            1910.91          Cash   
1              2      1860.0   1/1/2022 0:01           58590.27          Cash   
2              3     39158.0   1/1/2022 0:02             280.21    Debit Card   
3              4         NaN             NaN           86485.63           UPI   
4              5     12284.0   1/1/2022 0:04             385.89    Debit Card   

   Quantity  DiscountPercent     City StoreType  CustomerAge CustomerGender  \
0         1            35.00  Kolkata  In-Store         32.0           Male   
1         1             9.51  Lucknow    Online         40.0          Other   
2         1            16.55  Lucknow    Online         52.0          Other   
3         1             9.53   Jaipur  In-Store         62.0          Other   
4        22             0.98  Kolkata    Online         31.0         Female   

   LoyaltyPoints ProductName Region Returned  FeedbackScore  ShippingCost  \
0           2043     T-Shirt  North       No              4         24.31   
1           8490        Sofa   West       No              2       1502.96   
2           6328    Notebook  North      Yes              2         22.30   
3           1139      Laptop  North       No              4        461.38   
4             66       Apple  North       No              1         14.44   

   DeliveryTimeDays IsPromotional  
0                 5            No  
1                 7           Yes  
2                 2           Yes  
3                 7            No  
4                 1           Yes

In [ ]:
# Aggregate Insights
total_sales = df["TransactionAmount"].sum()
total_quantity = df["Quantity"].sum()
average_transaction = df["TransactionAmount"].mean()
payment_method_distribution = df["PaymentMethod"].value_counts()
store_type_distribution = df["StoreType"].value_counts()
average_discount = df["DiscountPercent"].mean()
average_shipping_cost = df["ShippingCost"].mean()
average_delivery_time = df["DeliveryTimeDays"].mean()

# Drill-down Insights
city_sales = df.groupby("City")["TransactionAmount"].sum().sort_values(ascending=False)
region_sales = df.groupby("Region")["TransactionAmount"].sum().sort_values(ascending=False)
product_sales = df.groupby("ProductName")["TransactionAmount"].sum().sort_values(ascending=False)
returned_products = df[df["Returned"] == "Yes"].groupby("ProductName")["TransactionAmount"].count()
promotional_sales = df.groupby("IsPromotional")["TransactionAmount"].sum()

# Customer Insights
age_distribution = df["CustomerAge"].describe()
gender_distribution = df["CustomerGender"].value_counts()

# Print results
print("### Aggregate Insights ###")
print(f"Total Sales Revenue: ₹{total_sales}")
print(f"Total Quantity Sold: {total_quantity}")
print(f"Average Transaction Amount: ₹{average_transaction}")
print("\nPayment Method Distribution:\n", payment_method_distribution)
print("\nStore Type Distribution:\n", store_type_distribution)
print(f"\nAverage Discount Percentage: {average_discount}%")
print(f"Average Shipping Cost: ₹{average_shipping_cost}")
print(f"Average Delivery Time: {average_delivery_time} days")

print("\n### Drill-Down Insights ###")
print("\nSales by City:\n", city_sales)
print("\nSales by Region:\n", region_sales)
print("\nSales by Product:\n", product_sales)
print("\nReturned Products Count:\n", returned_products)
print("\nSales by Promotion:\n", promotional_sales)

print("\n### Customer Insights ###")
print("\nCustomer Age Distribution:\n", age_distribution)
print("\nCustomer Gender Distribution:\n", gender_distribution)


### Aggregate Insights ###
Total Sales Revenue: ₹10202662960.19
Total Quantity Sold: 3747755
Average Transaction Amount: ₹20405.32592038

Payment Method Distribution:
 PaymentMethod
Debit Card     113015
Cash           112625
UPI            112517
Credit Card    111843
Name: count, dtype: int64

Store Type Distribution:
 StoreType
Online      225218
In-Store    224782
Name: count, dtype: int64

Average Discount Percentage: 24.999300279999993%
Average Shipping Cost: ₹397.30032198000004
Average Delivery Time: 5.239588 days

### Drill-Down Insights ###

Sales by City:
 City
Kolkata      1.027326e+09
Ahmedabad    1.023676e+09
Bangalore    1.022380e+09
Pune         1.022136e+09
Chennai      1.022123e+09
Delhi        1.021353e+09
Lucknow      1.021208e+09
Mumbai       1.018526e+09
Jaipur       1.015044e+09
Hyderabad    1.008892e+09
Name: TransactionAmount, dtype: float64

Sales by Region:
 Region
South    3.177273e+09
East     2.654969e+09
North    2.171503e+09
West     2.159912e+09
Name: Tr

# Key Insights from Sales Data

## Aggregate Insights:
- **Total Sales Revenue**: ₹10.2 billion — significant revenue across the dataset.
- **Total Quantity Sold**: 3,747,755 units — this indicates the total volume of products sold.
- **Average Transaction Amount**: ₹20,405 — a relatively high average transaction size.

### Payment Method Distribution:
- **Debit Card**: Most commonly used (113,015 transactions).
- **Cash**: Slightly less frequent (112,625 transactions).
- **UPI**: 112,517 transactions.
- **Credit Card**: 111,843 transactions.

### Store Type Distribution:
- **Online**: 225,218 transactions (slightly higher than In-Store).
- **In-Store**: 224,782 transactions (very close to Online).

### Discounts & Shipping:
- **Average Discount**: 25% — indicates heavy discounting on sales.
- **Average Shipping Cost**: ₹397 — represents the average cost of shipping items.
- **Average Delivery Time**: 5.24 days — typical shipping time for orders.

## Drill-Down Insights:

### Sales by City:
- **Top Cities**:
  - **Kolkata** (₹1.03 billion), **Ahmedabad** (₹1.02 billion), **Bangalore** (₹1.02 billion) — these cities lead in total sales.
  - Sales in major cities like **Delhi**, **Mumbai**, and **Chennai** also show high revenue.

### Sales by Region:
- **South**: ₹3.18 billion — leading region in sales.
- **East**: ₹2.65 billion — the second-highest in sales.
- **North**: ₹2.17 billion.
- **West**: ₹2.16 billion.

### Sales by Product:
- **Laptop**: ₹6.23 billion — the highest-grossing product.
- **Sofa**: ₹3.77 billion — significant sales, but far lower than laptops.
- **T-Shirt**: ₹102.3 million — much lower sales compared to other products.
- **Notebook**: ₹24 million — relatively small sales.
- **Apple**: ₹22.3 million — modest sales.

### Returned Products Count:
- **Apple**: 45,033 returned.
- **Laptop**: 44,904 returned.
- **Notebook**: 45,061 returned.
- **Sofa**: 44,696 returned.
- **T-Shirt**: 44,783 returned.
  - **Returned items** across all products seem quite similar, indicating potentially large transaction volumes and/or common return behaviors.

### Sales by Promotion:
- **Non-Promotional Sales**: ₹5.10 billion — almost equal to promotional sales.
- **Promotional Sales**: ₹5.09 billion — sales are nearly identical between promotional and non-promotional items, indicating that promotions are being utilized effectively but not overwhelmingly.

## Customer Insights:

### Customer Age Distribution:
- **Average Age**: 46 years, with a wide spread (18 to 74 years), suggesting diverse customer demographics.
- **Age Group Breakdown**:
  - **25%**: Customers aged 32 or younger.
  - **50%**: Customers aged 46 or younger.
  - **75%**: Customers aged 60 or younger.

### Customer Gender Distribution:
- **Other**: 150,257 (majority in the "Other" category).
- **Male**: 149,970 — almost equal to the number of female customers.
- **Female**: 149,773 — also almost equal to male customers.
  - The gender distribution is fairly balanced, with a higher proportion of customers labeled as "Other."

## Key Takeaways:
- **High-Value Products**: **Laptops** and **Sofas** are your top revenue drivers. You may want to focus more on promotions for these products.
- **Region-Specific Strategies**: The **South** region leads in sales, suggesting targeted strategies or inventory planning for this area.
- **Return Behavior**: Return rates seem consistent across product categories, which may warrant further analysis into return reasons or product quality.
- **Customer Base**: A majority of your customer base is older, with significant representation in the 32-60 age range, which could impact your marketing strategies.
- **Promotion Effectiveness**: Promotions don’t seem to drastically shift the revenue but still play a key role. Consider deeper analysis into which products benefit most from promotional campaigns.


# SQL Scripts for Data Exploration and Key Insights

## 1. Total Sales Revenue
```sql
SELECT SUM(TransactionAmount) AS total_sales
FROM sales_data;

2. Total Quantity Sold
SELECT SUM(Quantity) AS total_quantity
FROM sales_data;

3. Average Transaction Amount
SELECT AVG(TransactionAmount) AS avg_transaction
FROM sales_data;

4. Payment Method Distribution
SELECT PaymentMethod, COUNT(*) AS payment_method_count
FROM sales_data
GROUP BY PaymentMethod
ORDER BY payment_method_count DESC;

5. Store Type Distribution
SELECT StoreType, COUNT(*) AS store_type_count
FROM sales_data
GROUP BY StoreType
ORDER BY store_type_count DESC;

6. Average Discount Percentage
SELECT AVG(DiscountPercent) AS avg_discount
FROM sales_data;

7.Average Shipping Cost
SELECT AVG(ShippingCost) AS avg_shipping_cost
FROM sales_data;

8.Average Delivery Time
SELECT AVG(DeliveryTimeDays) AS avg_delivery_time
FROM sales_data;

9. Sales by City
SELECT City, SUM(TransactionAmount) AS city_sales
FROM sales_data
GROUP BY City
ORDER BY city_sales DESC;

10.Sales by Region
SELECT Region, SUM(TransactionAmount) AS region_sales
FROM sales_data
GROUP BY Region
ORDER BY region_sales DESC;

11.Sales by Product
SELECT ProductName, SUM(TransactionAmount) AS product_sales
FROM sales_data
GROUP BY ProductName
ORDER BY product_sales DESC;

12.Returned Products Count
SELECT ProductName, COUNT(*) AS returned_count
FROM sales_data
WHERE Returned = 'Yes'
GROUP BY ProductName
ORDER BY returned_count DESC;

13.Sales by Promotion
SELECT IsPromotional, SUM(TransactionAmount) AS promotional_sales
FROM sales_data
GROUP BY IsPromotional
ORDER BY promotional_sales DESC;

14.Customer Age Distribution (Basic Stats)
SELECT MIN(CustomerAge) AS min_age, MAX(CustomerAge) AS max_age, AVG(CustomerAge) AS avg_age,
       PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY CustomerAge) AS age_25_percentile,
       PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY CustomerAge) AS age_50_percentile,
       PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY CustomerAge) AS age_75_percentile
FROM sales_data;

15.Customer Gender Distribution
SELECT CustomerGender, COUNT(*) AS gender_count
FROM sales_data
GROUP BY CustomerGender
ORDER BY gender_count DESC;


# Key Insights
# 1. Total Sales Revenue: This gives you the total revenue from all transactions.
# 2. Total Quantity Sold: This tells you how many units were sold across all transactions.
# 3. Average Transaction Amount: This provides the average value of each transaction.
# 4. Payment Method Distribution: By counting the occurrences of each payment method, you can see which is the most common (e.g., Debit Card, Cash, etc.).
# 5. Store Type Distribution: Understanding the mix of Online and In-Store sales.
# 6. Average Discount Percentage: The average discount given on products.
# 7. Average Shipping Cost: Helps evaluate how much the company is spending on shipping per transaction.
# 8. Average Delivery Time: Insights into the typical shipping duration.
# 9. Sales by City: Shows which cities are generating the most sales.
# 10. Sales by Region: Identifies the regions with the highest sales performance.
# 11. Sales by Product: Helps pinpoint which products are driving the most sales.
# 12. Returned Products Count: Focuses on product returns to identify patterns or issues.
# 13. Sales by Promotion: Evaluates whether promotional sales are more profitable than non-promotional ones.
# 14. Customer Age Distribution: Provides demographic insights into the customer base by age.
# 15. Customer Gender Distribution: Reveals the gender breakdown of customers.
